### ISSUE TITLE
[R-package] lgb.prepare_rules() and lgb.prepare_rules2() should validate rules

### ISSUE THREAD

**Issue by jameslamb** [microsoft/LightGBM#2682]
## Summary

`lgb.prepare_rules() should validate that `rules` is a list of named numeric vectors.
`lgb.prepare_rules2()` should validate that `rules` is a list of named integer vectors.

"validate" = _throw an informative error_

## Motivation

Currently, `lgb.prepare_rules()` and `lgb.prepare_rules2()`  accept `rules` and apply them without validating the contents. This could silently cause undesirable effects.

The following code should result in an error because the values in `rules` are numeric, not integer

```r
testDF <- data.frame(
    character_col = c("a", NA_character_, "c", "a", "a", "c")
    , na_col = rep(NA, 6L)
    , na_real_col = rep(NA_real_, 6L)
    , na_int_col = rep(NA_integer_, 6L)
    , na_character_col = rep(NA_character_, 6L)
    , numeric_col = c(1.0, 9.0, NA_real_, 10.0, 11.0, 12.0)
    , factor_col = as.factor(c("n", "n", "y", "y", "n", "n"))
    , integer_col = c(1L, 9L, NA_integer_, 1L, 1L, 1L)
    , stringsAsFactors = FALSE
)
custom_rules <- list(
    "character_col" = c(
        "a" = 5.5
        , "c" = -10.2
    )
    , "factor_col" = c(
        "n" = 65.4
        , "y" = 65.01
    )
)
result <- lgb.prepare_rules2(
    data = input_data
    , rules = custom_rules
)
```

**Comment by jameslamb** [comment]
Closed in favor of being in #2302. We decided to keep all feature requests in one place.

Welcome to contribute this feature! Please re-open this issue (or post a comment if you are not a topic starter) if you are actively working on implementing this feature.

**Comment by jameslamb** [comment]
I've updated the title and description here now that #3095 has been merged
